File: php/Examples/V1/Buyers_Creatives/CreateHtmlCreatives.php
<?php

/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Google\Ads\AuthorizedBuyers\RealTimeBidding\Examples\V1\Buyers_Creatives;

use Google\Ads\AuthorizedBuyers\RealTimeBidding\ExampleUtil\BaseExample;
use Google\Ads\AuthorizedBuyers\RealTimeBidding\ExampleUtil\Config;
use Google_Service_RealTimeBidding_Creative;
use Google_Service_RealTimeBidding_HtmlContent;

/**
 * This example illustrates how to create HTML creatives for a given buyer account.
 */
class CreateHtmlCreatives extends BaseExample
{

    public function __construct($client)
    {
        $this->service = Config::getGoogleServiceRealTimeBidding($client);
    }

    /**
     * @see BaseExample::getInputParameters()
     */
    protected function getInputParameters()
    {
        return [
            [
                'name' => 'account_id',
                'display' => 'Account ID',
                'description' =>
                    'The resource ID of the buyers resource under which the creative is to be ' .
                    'created.',
                'required' => true
            ],
            [
                'name' => 'advertiser_name',
                'display' => 'Advertiser name',
                'description' => 'The name of the company being advertised in the creative.',
                'required' => false,
                'default' => 'Test'
            ],
            [
                'name' => 'creative_id',
                'display' => 'Creative ID',
                'description' =>
                    'The user-specified creative ID. The maximum length of the creative ID is ' .
                    '128 bytes',
                'required' => false,
                'default' => 'HTML_Creative_' . uniqid()
            ],
            [
                'name' => 'declared_attributes',
                'display' => 'Declared attributes',
                'description' =>
                    'The creative attributes being declared. Specify each attribute separated ' .
                    'by a comma.',
                'required' => false,
                'is_array' => true,
                'default' => ['CREATIVE_TYPE_HTML']
            ],
            [
                'name' => 'declared_click_urls',
                'display' => 'Declared click URLs',
                'description' =>
                    'The click-through URLs being declared. Specify each URL separated by a ' .
                    'comma.',
                'required' => false,
                'is_array' => true,
                'default' => ['http://test.com']
            ],
            [
                'name' => 'declared_restricted_categories',
                'display' => 'Declared restricted categories',
                'description' =>
                    'The restricted categories being declared. Specify each category separated ' .
                    'by a comma.',
                'required' => false,
                'is_array' => true
            ],
            [
                'name' => 'declared_vendor_ids',
                'display' => 'Declared vendor IDs',
                'description' =>
                    'The vendor IDs being declared. Specify each ID separated by a comma.',
                'required' => false,
                'is_array' => true
            ],
            [
                'name' => 'html_snippet',
                'display' => 'HTML snippet',
                'description' =>
                    'The HTML snippet that displays the ad when inserted in the web page.',
                'required' => false,
                'default' =>
                    '<iframe marginwidth=0 marginheight=0 height=600 frameborder=0 width=160 ' .
                    'scrolling=no src=\"https://test.com/ads?id=123456&curl=%%CLICK_URL_ESC%%' .
                    '&wprice=%%WINNING_PRICE_ESC%%\"></iframe>'
            ],
            [
                'name' => 'html_height',
                'display' => 'Height',
                'description' => 'The height of the HTML snippet in pixels.',
                'required' => false,
                'default' => 250
            ],
            [
                'name' => 'html_width',
                'display' => 'Width',
                'description' => 'The width of the HTML snippet in pixels.',
                'required' => false,
                'default' => 300
            ]
        ];
    }

    /**
     * @see BaseExample::run()
     */
    public function run()
    {
        $values = $this->formValues;
        $parentName = "buyers/$values[account_id]";

        $html = new Google_Service_RealTimeBidding_HtmlContent();
        $html->snippet = $values['html_snippet'];
        $html->height = $values['html_height'];
        $html->width = $values['html_width'];

        $newCreative = new Google_Service_RealTimeBidding_Creative();
        $newCreative->advertiserName = $values['advertiser_name'];
        $newCreative->creativeId = $values['creative_id'];
        $newCreative->declaredAttributes = $values['declared_attributes'];
        $newCreative->declaredClickThroughUrls = $values['declared_click_urls'];
        $newCreative->declaredRestrictedCategories = $values['declared_restricted_categories'];
        $newCreative->declaredVendorIds = $values['declared_vendor_ids'];
        $newCreative->setHtml($html);

        print "<h2>Creating Creative for '$parentName':</h2>";
        $result = $this->service->buyers_creatives->create($parentName, $newCreative);
        $this->printResult($result);
    }

    /**
     * @see BaseExample::getName()
     */
    public function getName()
    {
        return 'Create Buyer HTML Creative';
    }
}

File: java/src/main/java/com/google/api/services/samples/authorizedbuyers/realtimebidding/v1/buyers/creatives/CreateHtmlCreatives.java
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.api.services.samples.authorizedbuyers.realtimebidding.v1.buyers.creatives;

import com.google.api.services.realtimebidding.v1.RealTimeBidding;
import com.google.api.services.realtimebidding.v1.model.Creative;
import com.google.api.services.realtimebidding.v1.model.HtmlContent;
import com.google.api.services.samples.authorizedbuyers.realtimebidding.Utils;
import java.io.IOException;
import java.security.GeneralSecurityException;
import java.util.Collections;
import java.util.UUID;
import net.sourceforge.argparse4j.ArgumentParsers;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.ArgumentParserException;
import net.sourceforge.argparse4j.inf.Namespace;

/** Creates a creative with HTML content for the given buyer account ID. */
public class CreateHtmlCreatives {

  public static void execute(RealTimeBidding client, Namespace parsedArgs) throws IOException {
    Long accountId = parsedArgs.getLong("account_id");

    String parentBuyerName = String.format("buyers/%s", accountId);

    HtmlContent htmlContent = new HtmlContent();
    htmlContent.setSnippet(parsedArgs.getString("html_snippet"));
    htmlContent.setHeight(parsedArgs.getInt("html_height"));
    htmlContent.setWidth(parsedArgs.getInt("html_width"));

    Creative newCreative = new Creative();
    newCreative.setAdvertiserName(parsedArgs.getString("advertiser_name"));
    newCreative.setCreativeId(parsedArgs.getString("creative_id"));
    newCreative.setDeclaredAttributes(parsedArgs.<String>getList("declared_attributes"));
    newCreative.setDeclaredClickThroughUrls(parsedArgs.<String>getList("declared_click_urls"));
    newCreative.setDeclaredRestrictedCategories(
        parsedArgs.<String>getList("declared_restricted_categories"));
    newCreative.setDeclaredVendorIds(parsedArgs.<Integer>getList("declared_vendor_ids"));
    newCreative.setHtml(htmlContent);

    Creative creative = client.buyers().creatives().create(parentBuyerName, newCreative).execute();

    System.out.printf("Created creative for buyer Account ID '%s':\n", accountId);
    Utils.printCreative(creative);
  }

  public static void main(String[] args) {
    ArgumentParser parser =
        ArgumentParsers.newFor("CreateHtmlCreatives")
            .build()
            .defaultHelp(true)
            .description(("Creates an HTML creative for the given buyer account ID."));
    parser
        .addArgument("-a", "--account_id")
        .help("The resource ID of the buyers resource under which the creative is to be created. ")
        .required(true)
        .type(Long.class);
    parser
        .addArgument("--advertiser_name")
        .help("The name of the company being advertised in the creative.")
        .setDefault("Test");
    parser
        .addArgument("-c", "--creative_id")
        .help("The user-specified creative ID. The maximum length of the creative ID is 128 bytes.")
        .setDefault(String.format("HTML_Creative_%s", UUID.randomUUID()));
    parser
        .addArgument("--declared_attributes")
        .help(
            "The creative attributes being declared. Specify each attribute separated by a "
                + "space.")
        .nargs("*")
        .setDefault(Collections.singletonList("CREATIVE_TYPE_HTML"));
    parser
        .addArgument("--declared_click_urls")
        .help("The click-through URLs being declared. Specify each URL separated by a space.")
        .nargs("*")
        .setDefault(Collections.singletonList("http://test.com"));
    parser
        .addArgument("--declared_restricted_categories")
        .help(
            "The restricted categories being declared. Specify each category separated by a "
                + "space.")
        .nargs("*");
    parser
        .addArgument("--declared_vendor_ids")
        .help("The vendor IDs being declared. Specify each ID separated by a space.")
        .type(Integer.class)
        .nargs("*");
    parser
        .addArgument("--html_snippet")
        .help("The HTML snippet that displays the ad when inserted in the web page.")
        .setDefault(
            "<iframe marginwidth=0 marginheight=0 height=600 frameborder=0 width=160 "
                + "scrolling=no src=\"https://test.com/ads?id=123456&curl=%%CLICK_URL_ESC%%"
                + "&wprice=%%WINNING_PRICE_ESC%%\"></iframe>");
    parser
        .addArgument("--html_height")
        .help("The height of the HTML snippet in pixels.")
        .type(Integer.class)
        .setDefault(250);
    parser
        .addArgument("--html_width")
        .help("The width of the HTML snippet in pixels.")
        .type(Integer.class)
        .setDefault(300);

    Namespace parsedArgs = null;
    try {
      parsedArgs = parser.parseArgs(args);
    } catch (ArgumentParserException ex) {
      parser.handleError(ex);
      System.exit(1);
    }

    RealTimeBidding client = null;
    try {
      client = Utils.getRealTimeBiddingClient();
    } catch (IOException ex) {
      System.out.printf("Unable to create RealTimeBidding API service:\n%s", ex);
      System.out.println("Did you specify a valid path to a service account key file?");
      System.exit(1);
    } catch (GeneralSecurityException ex) {
      System.out.printf("Unable to establish secure HttpTransport:\n%s", ex);
      System.exit(1);
    }

    try {
      execute(client, parsedArgs);
    } catch (IOException ex) {
      System.out.printf("RealTimeBidding API returned error response:\n%s", ex);
      System.exit(1);
    }
  }
}
